bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java

```diff
@@ -3,34 +3,48 @@
 import javax.cache.CacheManager;
 import javax.cache.Caching;
 
-import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.autoconfigure.condition.*;
 import org.springframework.cache.jcache.JCacheCacheManager;
+import org.springframework.context.ApplicationEventPublisher;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
 import com.giffing.bucket4j.spring.boot.starter.config.cache.SyncCacheResolver;
 import com.giffing.bucket4j.spring.boot.starter.context.properties.Bucket4JBootProperties;
+import com.giffing.bucket4j.spring.boot.starter.context.properties.Bucket4JConfiguration;
 
 @Configuration
 @ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
-@ConditionalOnClass({ Caching.class, JCacheCacheManager.class })
+@ConditionalOnClass({Caching.class, JCacheCacheManager.class})
 @ConditionalOnBean(CacheManager.class)
-@ConditionalOnMissingBean(SyncCacheResolver.class)

```

Comment:
The reason for this check was to verify that this configuration is not applied when a different SynchCacheResolver is present to prevent two Configurations are activated at the same time. Is ist save to move it to the bean?

Reply:
This check is moved to line 32 above the method that creates the CacheResolver bean. This way the configuration class can also be used for creating the CacheManager and CacheListener beans, without being affected by an existing CacheResolver bean. The behaviour should still be the same, meaning that the CacheResolver bean will not be instantiated if one already exists.

Reply:
Isn't there a problem when to Caching Providers are in the classpath? For example JCache and Redis. If its the case, two of them are configured.

I long term solution would be to move each cache configuration to a separated Maven module, to make it explicit which cache configuration should be used and configured by this starter. But for now Its a problem. There was already someone in the past who has two cache implementation in the classpath with an configuration problem.

Reply:
The behaviour this should create is that the configuration class is always executed when (in case of JCache) a CacheManager bean exists, even if there already is a SyncCacheResolver. The method annotated with @Bean to create the JCacheCacheResolver does have the @ConditionalOnMissingBean annotation. This means that when someone creates his own instance of an AsyncCacheResolver, this bean will not created. The CacheManager and CacheListener for dynamically changing limits will however be made (if config caching is enabled).

Looking at it the way you said, where someone has multiple cache instances, it might be possible that 2 cache managers and listeners are created, So i can add the @ConditionalOnMissingBean annotation to those methods just to be sure. But the behaviour of creating the cache resolvers should still be functionally the same as how you implemented it.

Reply:
OK 👍